prov/util/src/util_srx.c

```diff
@@ -0,0 +1,1121 @@
+
+static struct util_rx_entry *util_init_unexp(struct util_srx_ctx *srx,
+			fi_addr_t addr, uint64_t size, uint64_t tag)
+{
+	struct util_rx_entry *util_entry;
+
+	util_entry = util_alloc_rx_entry(srx);
+	if (!util_entry)
+		return NULL;
+
+	util_entry->peer_entry.owner_context = NULL;
+	util_entry->peer_entry.size = size;
+	util_entry->peer_entry.addr = addr;
+	util_entry->peer_entry.tag = tag;
+
+	return util_entry;
+}
+
+static bool util_adjust_multi_recv(struct util_srx_ctx *srx,
+		struct fi_peer_rx_entry *rx_entry, size_t len)
+{
+	size_t left;
+	void *new_base;
+
+	left = rx_entry->iov[0].iov_len - len;
+
+	new_base = (void *) ((uintptr_t) rx_entry->iov[0].iov_base + len);
+	rx_entry->iov[0].iov_len = left;
+	rx_entry->iov[0].iov_base = new_base;
+	rx_entry->size = left;
+
+	return left < srx->min_multi_recv_size;
+}
+
+static struct util_rx_entry *util_process_multi_recv(struct util_srx_ctx *srx,
+		struct slist *queue, fi_addr_t addr, size_t size,
+		struct util_rx_entry *owner_entry)
+{
+	struct util_rx_entry *util_entry;
+
+	util_entry = util_get_recv_entry(srx,
+					 owner_entry->peer_entry.iov,
+					 owner_entry->peer_entry.desc,
+					 owner_entry->peer_entry.count, addr,
+					 owner_entry->peer_entry.context,
+					 owner_entry->peer_entry.tag,
+					 owner_entry->ignore,
+					 owner_entry->peer_entry.flags &
+					 (~FI_MULTI_RECV));
+	if (!util_entry)
+		return NULL;
+
+	if (util_adjust_multi_recv(srx, &owner_entry->peer_entry, size))
+		slist_remove_head(queue);
+
+	util_entry->peer_entry.owner_context = owner_entry;
+	owner_entry->multi_recv_ref++;
+
+	return util_entry;
+}
+
+static int util_match_msg(struct fid_peer_srx *srx, fi_addr_t addr, size_t size,
+			  struct fi_peer_rx_entry **rx_entry)
+{
+	struct util_srx_ctx *srx_ctx;
+	struct util_rx_entry *util_entry;
+	struct slist_entry *head;
+	int ret = FI_SUCCESS;
+
+	srx_ctx = srx->ep_fid.fid.context;
+	if (slist_empty(&srx_ctx->msg_queue)) {
+		util_entry = util_init_unexp(srx_ctx, addr, size, 0);
+		if (!util_entry)
+			return -FI_ENOMEM;
+		ret = -FI_ENOENT;
+	} else {
+		head = srx_ctx->msg_queue.head;
+		util_entry = container_of(head, struct util_rx_entry,
+					  peer_entry);
+		if (util_entry->peer_entry.flags & FI_MULTI_RECV) {
+			util_entry = util_process_multi_recv(srx_ctx,
+				&srx_ctx->msg_queue, addr, size, util_entry);
+			if (!util_entry) {
+				FI_WARN(&core_prov, FI_LOG_EP_CTRL,
+					"cannot allocate multi receive "
+					"buffer\n");
+				return -FI_ENOMEM;
+			}
+		} else {
+			(void) slist_remove_head(&srx_ctx->msg_queue);
+		}
+	}
+	util_entry->peer_entry.srx = srx;
+	*rx_entry = &util_entry->peer_entry;
+	return ret;
+}
+
+static int util_get_msg(struct fid_peer_srx *srx, fi_addr_t addr,
+		        size_t size, struct fi_peer_rx_entry **rx_entry)
+{
+	struct util_srx_ctx *srx_ctx;
+	struct util_rx_entry *util_entry, *any_entry;
+	struct slist *queue;
+	int ret;
+
+	srx_ctx = srx->ep_fid.fid.context;
+	srx_ctx->lock_acquire(&srx_ctx->lock);
```

Comment:
And I think there is a pitfall here... @wzamazon has pointed out in https://github.com/ofiwg/libfabric/commit/ab529c244a8ca74c0327e88c3baf1226e051f345 that `get_msg` and `queue_msg` must be serialized... which means you cannot release the lock after `get_msg` and before `queue_msg`, in case `generic_recv` will chime in and cause deadlock.

Example:
thread 1 call get_msg, did not find rx_entry in rx_list
thread 2 call fi_recv, put rx_entry to rx_list
thread 1 call queue_msg, put a different rx_entry to unexpected_rx_list

When thread 2 post the recv, it cannot find the unexp rx entry, because it was not queued yet. Then no thread can pick up that unexp rx entry

Reply:
@wenduwan The main locking change is [this](https://github.com/aingerson/libfabric/blob/peer/prov/util/src/util_srx.c#L1060-L1066) which turns the locking into a no-op. 

@shijin-aws Going to have to think about that example some more. I don't like the idea of not releasing the lock and relying on the peer to call queue properly. If it errors out during context allocation, we would need a different call just for the peer to release the lock without queueing. We will need some protection against that but I don't think playing with the locking is the safest choice. Let me weigh the options here a bit... Thanks for bringing this up.